How to Call Google APIs, first REST and RPC editions with Go, Swift, and Rust RPC samples.

timburks opened the pull request.

timburks committed:
New "How to Call Google APIs: RPC Edition" document.

timburks committed:
Add HowToREST and HowToRPCs documents and initial samples.

timburks committed:
Add Go sample, use GOOGLE_API_KEY environment variable for API keys.

timburks committed:
Add Go license header, add Swift sample.

timburks committed:
Add Rust sample.

wora commented on Examples/RPC/go/README.md:
```diff
@@ -0,0 +1,15 @@
+# Go HTTP RPC Sample
```
Please use lowercase and singular for all directory names. This is what we do for proto package names.

timburks replied:
Lower-casing "Examples", "RPC", and "REST" but keeping "examples" as plural.

wora commented on Examples/RPC/go/main.go:
```diff
@@ -0,0 +1,81 @@
+	if err != nil {
+		panic(err)
+	}
+	req.Header.Set("X-Goog-Api-Key", os.Getenv("GOOGLE_API_KEY"))
```
Please use a library. Otherwise the code is too messy.

timburks replied:
Thanks for your thoughts on this. I think our intent for these examples should be to make them as low-level as possible, particularly in terms of relying only on standard libraries. These would be the examples that people use to write their own client libraries. We can/will add higher-level examples elsewhere, maybe in language-specific repos.

wora commented on HowToRPC.md:
```diff
@@ -0,0 +1,64 @@
+- **Method.** This is the protobuf `rpc` name, such as "AnnotateText".
+
+### Requests
+RPC request messages are serialized and sent as the HTTP request body. For the server to handle the request properly, the client must set several HTTP request headers:
```
I strongly prefer 80-column or 100-column per line.

lukesneeringer replied:
Agreed, and 80, please.

timburks replied:
That's pretty horrible for editing. Do you use a text editor that automatically manages line breaks?

timburks replied:
I've auto-wrapped all Markdown using [pretty.sh](https://github.com/timburks/googleapis.github.io/blob/master/pretty.sh) (in the repo).

wora commented on HowToRPC.md:
```diff
@@ -0,0 +1,64 @@
+- **Service.** This is the fully qualified protobuf `service` name, such as "google.cloud.language.v1.LanguageService". In this case, “google.cloud.language.v1” is the package name in google/cloud/language/v1/language_service.proto and “LanguageService” is the name of the service section found on this line.
+- **Method.** This is the protobuf `rpc` name, such as "AnnotateText".
+
+### Requests
```
blank line before and after every heading.

timburks replied:
done

lukesneeringer commented on Examples/RPC/rust/RUNME:
```diff
@@ -0,0 +1,20 @@
+#!/bin/sh
```
Nit: Can this filename be `runme.sh`? (Not applicable if `RUNME` is a Rust convention I do not know about.)

timburks replied:
I'm changing this to SETUP.sh (I don't know if this is a general pattern, but I used this ("SETUP.sh") for iOS samples)

lukesneeringer commented on Examples/RPC/rust/src/main.rs:
```diff
@@ -0,0 +1,56 @@
+fn main() {
+    let google_api_key = env::var("GOOGLE_API_KEY").unwrap();
+
+    let mut document = language_service::Document::new();
```
I think this might be easier to understand if it either went completely "outside to inside" or "inside to outside". The current implementation mixes them. It starts with `Document`, _then_ sets its properties (the stuff inside it), then places it into the request (the stuff outside it).

It took me a couple reads to get a handle on this despite the fact that I knew the structure. I assert that an outside-in approach is easiest to follow, even if it results in more (or uglier) code.

timburks replied:
I've added a few blank lines to make the structure clearer. Otherwise I kept the structure as-is... I think there's always going to be some amount of "inside to out" since we are building a data structure and then converting it to binary (the final "outside" form). But let's discuss further (outside of this PR) and possibly revisit later.

lukesneeringer commented on Examples/RPC/go/README.md:
```diff
@@ -0,0 +1,15 @@
+# Go HTTP RPC Sample
+
+This directory contains a sample Go program that calls the Google Natural Language API using HTTP, RPC, and Protocol Buffers.
```
The phrase "HTTP, RPC, and protocol buffers" is weird to me. HTTP is a transport, protocol buffers is a wire format, and RPC is a....something. Technique?

If I approach this from the perspective of an unfamiliar face, I think I would be likely to think, "Okay, I know what HTTP is, but what is RPC and what are protocol buffers?" We can (and probably should) provide a link to the protobuf home page on each of these for the latter. For RPC, I think we either need something explaining what we mean here, or we should just remove it from _this_ list, even if we do want to explain the RPC technique elsewhere.

timburks replied:
Rephrasing this to "This directory contains a sample Go program that calls the Google Natural Language API using the gRPC Fallback protocol described in HowToRPC."

lukesneeringer commented on Examples/RPC/swift/Sources/main.swift:
```diff
@@ -0,0 +1,60 @@
+    exit(-1)
+}
+
+var document = Google_Cloud_Language_V1_Document()
```
This is obviously much easier to read (yay Swift) but my general feedback on the Rust example applies here. I think this would be easier to grok if it went fully "outside-in" or "inside-out"; mixing them is cognitively more challenging.

timburks replied:
Thanks, let's discuss outside the PR - as with the Rust example, I've added a few blank lines to hopefully make the structure clearer.

lukesneeringer commented on HowToREST.md:
```diff
@@ -0,0 +1,104 @@
+Google REST APIs are designed to be called by any standards-compliant HTTP client, including the [curl](https://curl.haxx.se/) command-line tool and library.
+
+### Authentication
+To use Google APIs, a client needs to use API keys for unauthenticated requests or OAuth tokens for authenticated requests. For more information, see [the Google Cloud Authentication Overview](https://cloud.google.com/docs/authentication/).
```
"use API keys for unauthenticated requests" is confusing. Fairly certain it is technically correct, but from a customer's perspective, they are authenticating _as themselves_ with their API key.

timburks replied:
Simplifying to "To use Google APIs, a client needs to authenticate with an API key or an OAuth token."

lukesneeringer commented on HowToREST.md:
```diff
@@ -0,0 +1,104 @@
+  -i
+HTTP/2 200 
+content-type: application/json; charset=UTF-8
+vary: X-Origin
```
I think it is reasonable and likely preferable to exclude everything from lines 41 to 51 in this diff, possibly with an "other headers" placeholder. These headers can and will change.

timburks replied:
SGTM. Removed.

lukesneeringer commented on HowToRPC.md:
```diff
@@ -0,0 +1,64 @@
+_For many developers, the easiest way to call a Google API is with one of our client libraries. But occasionally someone may prefer to make API calls directly - perhaps from a language or environment that we don’t support or using a different networking library or tool. Here we’ll show you how to do it._
+
+This page focuses on calling Google APIs directly using their underlying RPC interfaces. Most of these APIs are also available as REST services. For that, see [How to Call Google APIs, REST Edition](/HowToREST).
+
```
A paragraph saying "What is RPC?" would be valuable here.

timburks replied:
Adding links out to https://en.wikipedia.org/wiki/Remote_procedure_call for now.

lukesneeringer commented on HowToRPC.md:
```diff
@@ -0,0 +1,64 @@
+### Protocol Buffers 
+For all Google RPC APIs, the messages that are sent and received using the [Protocol Buffers](https://developers.google.com/protocol-buffers/docs/overview) encoding, and the definitive descriptions of these APIs are written in the Protocol Buffers Language. 
+
+To compile Protocol Buffer Language files, you’ll need `protoc`, the Protocol Buffer compiler. You can download `protoc` from the [google/protobuf release page](https://github.com/google/protobuf/releases) on GitHub or build it from source. You’ll probably also need a code generation plugin for the language that you’re using. Plugins are standalone executables written in many different languages, ...
```
Worth mentioning here that the releases page contains both distributions for language packages and protoc (broken up by architectures) in a single list. The person reading this wants the protoc for their architecture.

timburks replied:
Thanks, I've added a small clarification of this.

lukesneeringer commented on HowToRPC.md:
```diff
@@ -0,0 +1,64 @@
+
+To compile Protocol Buffer Language files, you’ll need `protoc`, the Protocol Buffer compiler. You can download `protoc` from the [google/protobuf release page](https://github.com/google/protobuf/releases) on GitHub or build it from source. You’ll probably also need a code generation plugin for the language that you’re using. Plugins are standalone executables written in many different languages, ...
+
+- **Go** [github.com/golang/protobuf](https://github.com/golang/protobuf)
```
Idea: Maybe we should link to the wiki page listing add-ons rather than have yet another mini list?
(I am not sure if this is actually what we should do, proposing for discussion.)

timburks replied:
I think initially it would be simple to keep all this together, but let's see how big this page gets.

lukesneeringer commented on HowToRPC.md:
```diff
@@ -0,0 +1,64 @@
+### A way to make API requests
+[gRPC](https://grpc.io/) is the recommended way to call Google RPC APIs. gRPC support is typically provided by additional `protoc` plugins that generate code for API clients and servers. This code uses lower-level primitives that send messages using gRPC’s HTTP/2-based messaging system, which supports request multiplexing, streaming APIs, and advanced flow control. To learn more about working wit ...
+
+If gRPC support is unavailable, Google APIs can also be called using HTTP/1 using the protocol described in the next section.
```
Should read just `HTTP`, not `HTTP/1`, yes?

lukesneeringer commented on HowToRPC.md:
````diff
@@ -0,0 +1,64 @@
+URL ::= BaseUrl "/" Service "/" Method
+```
+- **BaseUrl.** This is the base URL published by service owners, either via documentation or service discovery. For most Google APIs, the BaseUrl looks like "https://language.googleapis.com/$rpc". The base address can be found in the API reference documentation where it is identified as the “Service name”. For this example, “language.googleapis.com” is found in the Google Cloud Natural Language A ...
+- **Service.** This is the fully qualified protobuf `service` name, such as "google.cloud.language.v1.LanguageService". In this case, “google.cloud.language.v1” is the package name in google/cloud/language/v1/language_service.proto and “LanguageService” is the name of the service section found on this line.
````
Worth explicitly noting that the `.proto` file name is _not_ included here (it is based on the file's `package` declaration).

lukesneeringer replied:
Nit: use backticks, not quotes (and definitely not smart quotes).

timburks replied:
backticks - done, I'm not sure what you meant by the other comment (discuss later?)

lukesneeringer commented on HowToRPC.md:
```diff
@@ -0,0 +1,64 @@
+### Requests
+RPC request messages are serialized and sent as the HTTP request body. For the server to handle the request properly, the client must set several HTTP request headers:
+- **Content-Type.** The request message format is specified by the Content-Type header and must be "application/x-protobuf".
+X-Goog-Api-Key. This specifies a valid Google API key. It is optional if the Authorization header is used.
```
Needs bold facing to match the remainder of the list.

lukesneeringer commented on HowToRPC.md:
```diff
@@ -0,0 +1,64 @@
+
+### Requests
+RPC request messages are serialized and sent as the HTTP request body. For the server to handle the request properly, the client must set several HTTP request headers:
+- **Content-Type.** The request message format is specified by the Content-Type header and must be "application/x-protobuf".
```
`application/x-protobuf`

lukesneeringer commented on HowToRPC.md:
```diff
@@ -0,0 +1,64 @@
+The HTTP response contains at least the following headers:
+- **Content-Type.** This specifies the response serialization format. For normal responses and server errors, this will be "application/x-protobuf". Different values can be returned for network errors, such as when a message is rejected by a network proxy. All such errors will be accompanied by appropriate HTTP status codes.
+
+## [Examples](https://github.com/googleapis/googleapis.github.io/tree/master/Examples/RPC/)
```
Why are we not including the samples written for other languages? (Or is it on me to make a PR for my Python one?)

timburks replied:
I thought we could do a few first to establish a style, then invite more PRs. I believe @eoogbe already has a few more.

moorage commented on Examples/RPC/rust/README.md:
```diff
@@ -0,0 +1,17 @@
+
+## Protocol Buffer Support
+
+Message serialization requires Rust support code that is generated by protoc and the [Rust Protocol Buffers plugin](https://github.com/apple/swift-protobuf).
```
Probably not apple/swift-protobuf

timburks replied:
Yes, thanks for catching this :)

moorage commented on Examples/RPC/rust/README.md:
```diff
@@ -0,0 +1,17 @@
+
+## Instructions
+
+- Build and install `protoc-gen-rust` by running `cargo install protobuf` at the command line.
```
AFAICT, the current version of the protobuf crate does not come with `protoc-gen-rust`. Currently need `cargo install protobuf --vers 1.0.24`

timburks replied:
Thanks, I've noticed that this had changed. I've been using this to get the plugin:
`cargo install protobuf-codegen`

timburks committed:
Convert Examples, RPC, and REST directory names to lower-case.

timburks committed:
Wrap all Markdown to 80 columns.

Text is wrapped with prettier.
% prettier --write --print-width 80 --prose-wrap always <filename.md>
Run pretty.sh at the root of this repo to "prettier" all Markdown files.

timburks committed:
Rename RUNME files to SETUP.sh and add license headers.

timburks commented on Examples/RPC/swift/RUNME:
```diff
@@ -0,0 +1,22 @@
+#!/bin/sh
```
SETUP.sh?

timburks committed:
Remove unnecessary headers from HowToREST sample.

timburks committed:
Replace quotes with backticks.

timburks committed:
Suggested edits to "HowTo" docs plus explicit mention of the gRPC fallback protocol.

timburks committed:
Add links to Wikipedia description of RPC, mention gRPC Fallback explicitly in example READMEs.

timburks committed:
Add (minimal) background detail on protoc.

timburks committed:
update and verify go/swift/rust samples.

wora commented on HowToRPC.md:
```diff
@@ -0,0 +1,166 @@
+# How to Call Google APIs: RPC Edition
+
+_For many developers, the easiest way to call a Google API is with one of our
+client libraries. But occasionally someone may prefer to make API calls
```
someone => one

timburks replied:
Could you explain why you prefer "one"? 

Reviewing the comments and the answer to [this question comparing one vs. someone](https://english.stackexchange.com/questions/120592/one-vs-someone-can-be-used-interchangeably), I think "someone" is a better fit: by indicating a "generic specific" individual, it implies that directly making API calls is a less-common situation. 

If we said "one" it would seem like we might be [circumspectly/euphemistically referring to ourselves or the reader](https://english.stackexchange.com/questions/120592/one-vs-someone-can-be-used-interchangeably#comment247695_120592), which would seem like a hint that this is actually a preferred practice and not a fallback.

lukesneeringer replied:
Considering we use "you" two lines later, it is fine to use it here also.

wora commented on HowToRPC.md:
```diff
@@ -0,0 +1,166 @@
+a different networking library or tool. Here we’ll show you how to do it._
+
+This page focuses on calling Google APIs directly using their underlying RPC
+([Remote Procedure Call](https://en.wikipedia.org/wiki/Remote_procedure_call))
```
[RPC](...)

People rarely use the term remote procedure call these days, so it is not useful to spell it out.

timburks replied:
@lukesneeringer had asked for this in an earlier comment:

| A paragraph saying "What is RPC?" would be valuable here.

I think spelling it out and linking to wikipedia is a very small (minimal?) way to provide that, so I'd prefer to keep this.

wora replied:
Google search shows 35.5M use of RPC, only 824k use of "remote procedure call". Most people in the world don't speak native English. There is no point to force them to read more English words, like spelling out HTTP and RPC.

wora commented on HowToRPC.md:
```diff
@@ -0,0 +1,166 @@
+
+This page focuses on calling Google APIs directly using their underlying RPC
+([Remote Procedure Call](https://en.wikipedia.org/wiki/Remote_procedure_call))
+interfaces. Most of these APIs are also available as REST services. For that,
```
Most of these APIs => Most Google APIs

timburks replied:
OK, will change.

wora commented on HowToRPC.md:
```diff
@@ -0,0 +1,166 @@
+
+- **Authorization.** This specifies a valid Google OAuth access token in the
+  format of `Bearer {token}`. It is optional if the request is unauthenticated
+  or if an API key is used.
```
unauthenticated and an API key is provided.

timburks replied:
I believe @lukesneeringer felt that "unauthenticated" would be unclear to some readers. 

I'll change this to "It is optional if an API key is provided."

wora commented on HowToRPC.md:
```diff
@@ -0,0 +1,166 @@
+
+- **User-Agent.** This should contain a meaningful string that identifies the
+  client application for analytics and troubleshooting purposes. It is usually
+  provided automatically by HTTP support libraries.
```
Except browser, User-Agent is not provided by the library.

timburks replied:
I'm removing User-Agent from this list.

wora commented on HowToRPC.md:
```diff
@@ -0,0 +1,166 @@
+For successful requests, the HTTP status code is `200` and the HTTP response
+body contains the serialized RPC response message. For unsuccessful requests,
+the HTTP status code is the HTTP mapping for `google.rpc.Code` and the HTTP
+response body contains a serialized `google.rpc.Status` message.
```
Provide a link to api style guide, errors section.

timburks replied:
Done.

timburks committed:
Text edits suggested in review by wora.

timburks committed:
Mark gRPC Fallback as experimental; url format subject to change.

wora commented on HowToRPC.md:
```diff
@@ -0,0 +1,166 @@
+- **Swift**
+  [github.com/apple/swift-protobuf](https://github.com/apple/swift-protobuf)
+
+### A way to make API requests
```
Making API requests

timburks replied:
OK, will change.

wora commented on HowToRPC.md:
```diff
@@ -0,0 +1,166 @@
+  by a network proxy. All such errors will be accompanied by appropriate HTTP
+  status codes.
+
+## [Examples](https://github.com/googleapis/googleapis.github.io/tree/master/examples/rpc/)
```
Google APIs don't use plural words for directory name. I prefer we follow the tradition here.

timburks replied:
Is there a style guide or example of that? When I look at the [API Style Guide Naming Conventions](https://cloud.google.com/apis/design/naming_convention), I find

> Collection IDs should use plural form and lowerCamelCase, and American English spelling and semantics. For example: events, children, or deletedEvents.

Isn't this directory a collection of examples?

wora commented on HowToREST.md:
```diff
@@ -0,0 +1,135 @@
+> including C, C++, C#, Java, JavaScript, Perl, Python, and many others. These
+> properties make JSON an ideal data-interchange language.
+
+### A way to make API requests
```
This title is not very useful for search optimization. How about "Making REST API requests" or "Calling Google REST APIs"

timburks replied:
OK, will change.

timburks committed:
Edits requested by wora.

lukesneeringer commented on HowToREST.md:
```diff
@@ -0,0 +1,135 @@
+# How to Call Google APIs: REST Edition
+
+_For many developers, the easiest way to call a Google API is with one of our
+client libraries. But occasionally someone may prefer to make API calls
```
Perhaps add a link to https://cloud.google.com/apis/docs/cloud-client-libraries?

timburks replied:
adding, thanks!

lukesneeringer commented on HowToRPC.md:
```diff
@@ -0,0 +1,165 @@
+
+Most Google APIs are designed as
+[RPC](https://en.wikipedia.org/wiki/Remote_procedure_call) APIs using the
+[Protocol Buffers language](https://developers.google.com/protocol-buffers/docs/overview).
```
I would substitute "the protocol buffers language" for just "protocol buffers".

timburks replied:
OK

lukesneeringer commented on HowToRPC.md:
```diff
@@ -0,0 +1,165 @@
+and the plugin interface is defined in the
+[plugin.proto](https://github.com/google/protobuf/blob/master/src/google/protobuf/compiler/plugin.proto)
+file. Here are some plugins that we have used:
+
```
It might be worth listing the languages that ship with protoc (which is C++, Obj-C, Java, Java Lite, Python, Ruby, and PHP, I _think_).

timburks replied:
Adding a link to https://developers.google.com/protocol-buffers/docs/reference/overview

lukesneeringer commented on HowToRPC.md:
```diff
@@ -0,0 +1,165 @@
+and libraries. For a sample command-line client, see the
+[oauth2l](https://github.com/google/oauth2l) on GitHub.
+
+## gRPC Fallback (Experimental)
```
This is out of experimental now, yes?

timburks replied:
almost :)

lukesneeringer reviewed (approved):
Approved, consider all comments to be suggestions only and not blockers.

timburks committed:
Edits suggested by lukesneeringer.

mCloudCorp-sudo commented on CONTRIBUTING.md:
```diff
@@ -20,4 +20,4 @@ again.
 All submissions, including submissions by project members, require review. We
 use GitHub pull requests for this purpose. Consult
 [GitHub Help](https://help.github.com/articles/about-pull-requests/) for more
-information on using pull requests.
```
Mccoy.stevens@lexisnexisrisksolutions.net

mCloudCorp-sudo commented on HowToREST.md:
```diff
@@ -0,0 +1,135 @@
+# How to Call Google APIs: REST Edition
```
Mccoy.stevens@lexisnexisrisksolutions.net 
https://github.com/googleapis/googleapis.github.io/blob/license/examples/rest/CURL.sh

mCloudCorp-sudo commented on examples/rpc/go/main.go:
```diff
@@ -0,0 +1,81 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
```
Mccoy.stevens@lexisnexisrisksolutions.net

mCloudCorp-sudo commented on HowToRPC.md:
```diff
@@ -0,0 +1,169 @@
+# How to Call Google APIs: RPC Edition
```
Mccoy.stevens@lexisnexisrisksolutions.net

mCloudCorp-sudo commented on examples/rpc/rust/README.md:
```diff
@@ -0,0 +1,34 @@
+# Rust HTTP RPC Sample
+
```
Mccoy.stevens@lexisnexisrisksolutions.net

mCloudCorp-sudo commented on examples/rpc/rust/SETUP.sh:
```diff
@@ -0,0 +1,36 @@
+#!/bin/sh
+#
+# Copyright 2018 Google Inc. All Rights Reserved.
```
Mccoy.stevens@lexisnexisrisksolutions.net

mCloudCorp-sudo commented on examples/rpc/rust/src/main.rs:
```diff
@@ -0,0 +1,58 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
```
Mccoy.stevens@lexisnexisrisksolutions.net

mCloudCorp-sudo commented on examples/rpc/swift/README.md:
```diff
@@ -0,0 +1,30 @@
+# Swift HTTP RPC Sample
+
```
Mccoy.stevens@lexisnexisrisksolutions.net
